src/AbstractMenuController.cpp

```diff
@@ -0,0 +1,167 @@
+/*
+ * Memoris
+ * Copyright (C) 2015  Jean LELIEVRE
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/**
+ * @file AbstractMenuController.cpp
+ * @package controllers
+ * @author Jean LELIEVRE <Jean.LELIEVRE@supinfo.com>
+ */
+
+#include "AbstractMenuController.hpp"
+
+namespace memoris
+{
+namespace controllers
+{
+
+/**
+ *
+ */
+AbstractMenuController::AbstractMenuController(utils::Context& context) :
+    Controller(context)
+{
+    /* the default pointed item is always the first one when the menu starts */
+    selectorPosition = 0;
+}
+
+/**
+ *
+ */
+AbstractMenuController::~AbstractMenuController()
+{
+    /* empty destructor, must be defined to make the class abstract */
+}
+
+/**
+ *
+ */
+void AbstractMenuController::addMenuItem(items::MenuItem* item)
+{
+    items.push_back(item);
+}
+
+/**
+ *
+ */
+void AbstractMenuController::renderAllMenuItems(utils::Context& context)
+{
+    for(items::MenuItem* item : items)
+    {
+        item->render(context);
+    }
+}
+
+/**
+ *
+ */
+unsigned short AbstractMenuController::getLastMenuItemIndex() const
+{
+    /* convert the size_t to unsigned short for future usage; unsigned
+       short has usually a maximum value of 31767; we use this function
+       for menu items; we can considere this method as a safe method */
+    unsigned short size = static_cast<unsigned short>(items.size());
+
+    /* substract one if the size is more than 0, to get the last item index */
+    if (size > 0)
+    {
+        return size - 1;
+    }
+
+    return size;
+}
+
+/**
+ *
+ */
+void AbstractMenuController::updateMenuSelection()
+{
+    /* browse all the menu items; use this way to keep trace of the
+       current browsed item index */
+    for(
+        std::vector<items::MenuItem*>::iterator iterator = items.begin();
+        iterator != items.end();
+        ++iterator
+    )
+    {
+        /* select the item that has the index equals to the selector
+           position; use a cast to force std::distance to always return
+           an unsigned short for the comparison */
+        if (
+            static_cast<unsigned short>(
+                std::distance(
```

Comment:
wrap into std::ads for more security : add comment that std::distance can return a signed value; keep the cast for explanation